so-staus detects tty

JamesMConroy opened the pull request:
Resolves #6495 

I changed `so-status` print colors if and only if the output is a tty.

Also removed a redundant function and refactored how the line was printed

JamesMConroy committed:
Updated the useage function to use printf

Using a hear doc means we have to exactly specify the formatting. Useing
printf handles formatting for us

JamesMConroy committed:
Only print color codes if we're printing to a tty

If we're not printing to a tty the escape sequences can only clutter the
screen.

Also removed a redundant function to print lines if not printing to a
tty. It was only called if docker wasn't running, not if the output
wasn't a tty.

JamesMConroy committed:
Print help message if not running as root

JamesMConroy committed:
Refactored to reduce length and number of lines

JamesMConroy committed:
Reworked tty detection and status printing

I was able to reduce the line count and make the script more reliable

JamesMConroy committed:
Added Comments for future enhancements

JamesMConroy committed:
Moved line declaration out of tty conditional

This way it will always be set to ""

TOoSmOotH commented:
We will take a deep look at this once cases comes out.

rwwiv commented on salt/common/tools/sbin/so-status:
```diff
@@ -170,78 +176,70 @@ parse_status() {
-
-    if [[ $service_state = "$ERROR_STRING" ]] || [[ $service_state = "$MISSING_STRING" ]]; then
-        if [[ "$EXITCODE" -eq 0 ]]; then
-            EXITCODE=1
-        fi
-    elif [[ $service_state = "$PENDING_STRING" ]] || [[ $service_state = "$DISABLED_STRING" ]] || [[ $service_state = "$STARTING_STRING" ]] || [[ $service_state = "$WAIT_START_STRING" ]]; then
-        EXITCODE=2
+    if [[ $service_state = "$ERROR_STRING" ]] \
+        || [[ $service_state = "$MISSING_STRING" ]]; then
+          state_color="${red:-}"
+          if [[ "$EXITCODE" -eq 0 ]]; then
+              EXITCODE=1
+          fi
+    elif [[ $service_state = "$SUCCESS_STRING" ]]; then
+        state_color="${green:-}"
+    elif [[ $service_state = "$PENDING_STRING" ]] \
+        || [[ $service_state = "$DISABLED_STRING" ]] \
+        || [[ $service_state = "$STARTING_STRING" ]] \
+        || [[ $service_state = "$WAIT_START_STRING" ]]; then
+            state_color="${yellow:-}"
+            EXITCODE=2
     fi
 
-    printf "    $service_name "
-    for i in $(seq 0 $(( 35 - ${#service_name} - ${#service_state} ))); do
-        printf "-"
-    done
-    printf " [ "
-    printf "$service_state"
-    printf "%s    \n" " ]"
+    service_state="${state_color:-}${service_state}${NC:-}"
+    line="${state_color:-}${line:-}${NC:-}"
+    printf "\t%s %s [ %s ]\n" "${service_name}" "${line:-}" "${service_state}"
```
Much easier to read this code, this makes the function much simpler to maintain.

rwwiv commented on salt/common/tools/sbin/so-status:
```diff
@@ -257,9 +255,13 @@ main() {
-          local focus_color="\e[1;34m"
-          printf "\n"
-          printf "${focus_color}%b\e[0m" "Checking Docker status\n\n"
+          if (( __tty == 1 ))
+          then
```
Use `if ... ; then` syntax rather than adding a newline.

rwwiv commented on salt/common/tools/sbin/so-status:
```diff
@@ -15,21 +15,23 @@
-   echo "This command must be run as root"
-   exit 1
-fi
+MYNAME="${0##*/}"
```
This breaks standard from the other Security Onion scripts, which use the full path (i.e. `$0`).

JamesMConroy replied:
I'd still like to capture the value of `$0` at script start since there's no guarantee that `$0` wont change. We can still reference the scrip's full path with:

```
$(readlink -f ${MYPATH}/${MYNAME})
```

Or something like that.

rwwiv replied:
This script isn't doing anything that might change `$0`, and the value is being used solely by the usage function `display_help()` and the root check. There's no need to capture its value.

However,  I was conflating the full path and the value of `$0` so referencing the full path isn't actually necessary either.

JamesMConroy replied:
The value of `$0` is shell dependent so silly things can happen, but it's been changed back.

rwwiv replied:
In this case we don't need to worry about other shells since the shebang specifies this runs under bash.

rwwiv commented on salt/common/tools/sbin/so-status:
```diff
@@ -290,22 +291,29 @@ main() {
+    return "${__tty}"
+}
+
+if ! [ "$(id -u)" = 0 ]; then
+   echo "${MYNAME}: This command must be run as root"
+   display_help 1
+fi
```
This is included at the top of the script so nothing else is evaluated before checking for root, and displaying the usage message here also breaks from the standard set by other scripts in Security Onion.

JamesMConroy replied:
But what if all I want to do when running is to check the script's usage? Should you really need root just to see the usage?

There's probably a better way to handle that use case, while staying consistent with the rest of the scripts.

rwwiv replied:
I think there's a valid argument to be made for parsing arguments before checking root and showing usage if that flag is passed in, but any change like that should be made to all Security Onion scripts to maintain consistency.

So for now, let's revert this change. But feel free to [start a discussion](https://github.com/Security-Onion-Solutions/securityonion/discussions/new) to put forward the idea of handling this differently and the team can look it over.

JamesMConroy replied:
Does this scrip even need to run as root?

One nice consequence of using the data from the socket is that the calling user doesn't even need to be in the docker group.

JamesMConroy replied:
To test this script as a regular user I removed the check completely.

rwwiv replied:
If you notice, these scripts are all included in `/usr/sbin`, so by convention (and by design) they require root.

JamesMConroy replied:
Yea I'll probably start a discussion about it. That whole principal of least privilege and everything.

JamesMConroy replied:
Discussion made https://github.com/Security-Onion-Solutions/securityonion/discussions/7156

rwwiv commented on salt/common/tools/sbin/so-status:
```diff
@@ -290,22 +291,29 @@ main() {
+    [ -t 1 ] \
+        && __tty=1
+    # don't print colors if NO_COLOR is set to anything
+    [ "${#NO_COLOR}" -ne 0 ] \
+        && __tty=0
```
Why do you add a line break here and on L296-297?

JamesMConroy replied:
Personal style/preference. I find it easier to read as two lines as opposed to just one.

rwwiv replied:
Let's stay consistent with the rest of Security Onion and keep these on one line.

rwwiv commented on salt/common/tools/sbin/so-status:
```diff
@@ -290,22 +291,29 @@ main() {
+    # don't print colors if NO_COLOR is set to anything
+    [ "${#NO_COLOR}" -ne 0 ] \
+        && __tty=0
+    return "${__tty}"
```
There's no need to return if you aren't doing anything with the return value.

JamesMConroy replied:
Fixed

rwwiv commented on salt/common/tools/sbin/so-status:
```diff
@@ -290,22 +291,29 @@ main() {
-      display_help
-      exit 0
-      ;;
+    h) display_help 0 ;;
+    q) QUIET=true ;;
+    \?) display_help 1 ;;
```
As above, the usage/help function shouldn't be handling exiting the script.

rwwiv commented on salt/common/tools/sbin/so-status:
```diff
@@ -101,9 +103,13 @@ create_expected_container_list() {
 }
 
 populate_container_lists() {
+    # TODO: check exit code directly, not with $?
```
This was for legibility but could just as easily be written to check the exit code directly.

rwwiv reviewed (changes requested):
This PR needs some attention before we can approve it. Additionally, don't change functionality outside of what your PR is trying to accomplish.

JamesMConroy commented on salt/common/tools/sbin/so-status:
```diff
@@ -15,21 +15,23 @@
-	              single status code for overall status
-	0:Ok, 1:Error, 2:Starting/Pending, 99:Installing SO
-HELP_USAGE
+  printf "%s" "${MYNAME}"
+  printf " [%s]" "-h" "-q|--quiet"
+  printf "\n"
+  printf "\t-%s\t%s\n" \
+    "h" "Show this message." \
+    "q,--quiet" "suppress the output and return a single status code of overall status"
+  printf "\nReturn Codes:\n"
+  printf "\t%s - %s\n" \
+  "0" "Ok" \
+  "1" "Error" \
+  "2" "Starting/Pending"\
+  "99" "Installing SO"
```
Yea mostly that, but this way I don't have to format the message myself and the here doc doesn't mess up my editors auto indent/

rwwiv replied:
Let's keep this consistent with the rest of the project then and revert the change.

JamesMConroy commented on salt/common/tools/sbin/so-status:
```diff
@@ -15,21 +15,23 @@
+  "1" "Error" \
+  "2" "Starting/Pending"\
+  "99" "Installing SO"
+  # exit with code 0, unless passed a number when called
+  exit "${1:-0}"
```
I disagree, it is incredibly useful to have a consistent way to print the usage and exit the script.

rwwiv replied:
There already exists a consistent way to print usage and exit:
```bash
display_help
exit 0

# or

display_help; exit 0
```

This adds functionality that isn't readily apparent by the name of the function and changes its behavior for the sake of not typing `exit`

I think it probably does need to be renamed to `usage` to match the rest of the scripts, but even then the function should only perform one action.

JamesMConroy replied:
I disagree strongly, but you're the ones who have to maintain this scrip.

rwwiv replied:
This isn't a matter of opinion, just good programming practice.

A good rule of thumb is that if your function name needs to include `and` to describe what it's doing, you should split it up. 
e.g. `display_help_and_exit()`.

JamesMConroy commented on salt/common/tools/sbin/so-status:
```diff
@@ -101,9 +103,13 @@ create_expected_container_list() {
     systemctl is-active --quiet docker
 
     if [[ $? = 0 ]]; then
+        # TODO: why use curl when you can just run docker ps?
```
I will update the TODO to reflect that.

JamesMConroy commented on salt/common/tools/sbin/so-status:
```diff
@@ -101,9 +103,13 @@ create_expected_container_list() {
     if [[ $? = 0 ]]; then
+        # TODO: why use curl when you can just run docker ps?
+        # TODO: why use an indexed array when bash has associtive arrays?
+        # TODO: why use an array when you can just use a string?
```
That's mostly an artifact of me working on a **POSIX** shell only shop so I tend to look at the world in terms of strings only.

JamesMConroy commented on salt/common/tools/sbin/so-status:
```diff
@@ -170,78 +176,70 @@ parse_status() {
+    local service_state="" ; service_state="$( parse_status "${1}" "${2}" )"
+    # XXX: What will we do if tput isn't avalable?
+    local line=""
+    if (( __tty == 1 )); then
+        local NC; NC="$(tput sgr0)" # no color
+        local red=""; red="$(tput setaf 1 bold)"
+        local green; green="$(tput setaf 2 bold)"
+        local yellow; yellow="$(tput setaf 3 bold)"
+
+        # construct a line of '------' so that the names and states are all aligned
+        local PADDING_CONSTANT=16 # two tabs wide
+        PADDING_CONSTANT=$((PADDING_CONSTANT + PADDING_CONSTANT / 2))
+        local columns ; columns=$(tput cols)
+        linewidth=$(( columns - PADDING_CONSTANT - ${#service_name} - ${#service_state} ))
+
+        for i in $(seq 0 "${linewidth}"); do
+          line="${line}-"
+        done
```
Yes, I'm not sure there is a way to get the number of columns if you're not printing to a tty, and if I'm parsing output with sed/awk I don't really want a long decorative line.

rwwiv replied:
I'm going to talk to the team about this one. I think you make a great point, that removing them when not in a tty makes parsing the output programmatically much easier.

To your point about getting the number of columns, it looks like the script previously just hardcoded the width to 35 characters total.

rwwiv replied:
Looks like the team would prefer to keep the lines in all output. If you wanted to parse the output with sed/awk you could use `tr` to remove the dash character.

JamesMConroy replied:
Alright the line prints all the time now.

JamesMConroy committed:
Added Changes Suggested by Reviewer

Added a missing semi colon between a local variable's declaration and
assignment
Removed an unused return value
Made a TODO more descriptive

JamesMConroy commented on salt/common/tools/sbin/so-status:
```diff
@@ -257,9 +255,13 @@ main() {
-          printf "${focus_color}%b\e[0m" "Checking Docker status\n\n"
+          if (( __tty == 1 ))
+          then
+          local focus_color="" focus_color="$(tput setaf 3 bold)"
```
Fixed

JamesMConroy committed:
Revert improvements to usage function

Made to make it more consistent with the rest of the scripts in
Security Onion

JamesMConroy committed:
Removed MYNAME variable

Preferring to just use the value of $0 instead

JamesMConroy committed:
Always print a line of '-'

Even when not printing to a tty

This is behavior preferred by the team

JamesMConroy committed:
Don't split lines after standalone tests

This is to make the formatting consistent with the rest of the scripts

rwwiv reviewed (changes requested):
Please ensure you're preserving the output of the script, I've just now compared the non-tty and tty outputs and neither match the original.

rwwiv commented on salt/common/tools/sbin/so-status:
```diff
@@ -290,22 +285,31 @@ main() {
+    [ "${#NO_COLOR}" -ne 0 ] && __tty=0
+}
+
+if ! [ "$(id -u)" = 0 ]; then
+   echo "${0}: This command must be run as root"
+fi
```
This should also exit the script with a nonzero exit code.

JamesMConroy replied:
done

JamesMConroy committed:
Don't set the padding constant if not in a tty

This will preserve the original width from before my changes

JamesMConroy commented on salt/common/tools/sbin/so-status:
```diff
@@ -170,78 +170,71 @@ parse_status() {
-    printf "${state_color}%b\e[0m" "$service_state"
-    printf "%s    \n" " ]"
-}
-
-non_term_print_line() {
-    local service_name=${1}
-    local service_state="$( parse_status ${1} ${2} )"
 
-    if [[ $service_state = "$ERROR_STRING" ]] || [[ $service_state = "$MISSING_STRING" ]]; then
-        if [[ "$EXITCODE" -eq 0 ]]; then
-            EXITCODE=1
-        fi
-    elif [[ $service_state = "$PENDING_STRING" ]] || [[ $service_state = "$DISABLED_STRING" ]] || [[ $service_state = "$STARTING_STRING" ]] || [[ $service_state = "$WAIT_START_STRING" ]]; then
-        EXITCODE=2
+    if [[ $service_state = "$ERROR_STRING" ]] \
+        || [[ $service_state = "$MISSING_STRING" ]]; then
+          state_color="${red:-}"
+          if [[ "$EXITCODE" -eq 0 ]]; then
+              EXITCODE=1
+          fi
+    elif [[ $service_state = "$SUCCESS_STRING" ]]; then
+        state_color="${green:-}"
+    elif [[ $service_state = "$PENDING_STRING" ]] \
+        || [[ $service_state = "$DISABLED_STRING" ]] \
+        || [[ $service_state = "$STARTING_STRING" ]] \
+        || [[ $service_state = "$WAIT_START_STRING" ]]; then
+            state_color="${yellow:-}"
+            EXITCODE=2
     fi
 
-    printf "    $service_name "
-    for i in $(seq 0 $(( 35 - ${#service_name} - ${#service_state} ))); do
-        printf "-"
-    done
-    printf " [ "
-    printf "$service_state"
-    printf "%s    \n" " ]"
+    service_state="${state_color:-}${service_state}${NC:-}"
+    line="${state_color:-}${line:-}${NC:-}"
+    printf "\t%s %s [ %s ]\n" "${service_name}" "${line:-}" "${service_state}"
```
ahh the check tty was only run in the print line function so it wasn't set when printing the 'Checking * status' lines

JamesMConroy committed:
Check for tty in main

So that the value is set every time it is checked

JamesMConroy committed:
Exit with an error code if the user isn't root

JamesMConroy commented on salt/common/tools/sbin/so-status:
```diff
@@ -170,78 +170,71 @@ parse_status() {
-    for i in $(seq 0 $(( $columns - $PADDING_CONSTANT - ${#service_name} - ${#service_state} ))); do
-        printf "${state_color}%b\e[0m" "-"
+    # construct a line of '------' so that the names and states are all aligned
+    linewidth=$(( columns - PADDING_CONSTANT - ${#service_name} - ${#service_state} ))
```
hopefully that fixes the width issue

rwwiv reviewed (changes requested):
I'm still not seeing the correct output with your changes. Did you compare the output to the image I included in my previous comment? I've suggested several changes that should bring this PR to the correct state.

PR:

<img width="644" alt="PR (tty)" src="https://user-images.githubusercontent.com/4155337/153448190-9cfd941e-1ed1-4976-826a-13b991dd1e64.png">

Original:

<img width="644" alt="original (tty)" src="https://user-images.githubusercontent.com/4155337/153448230-e422f266-d078-48ab-8d12-4a07a1795547.png">

PR with changes:

<img width="644" alt="PR (tty with changes)" src="https://user-images.githubusercontent.com/4155337/153457912-2bde3a67-0b5e-475b-8ed8-a9e2a44c1de7.png">

JamesMConroy committed:
Seperate bold attribute from colors

As suggested by @rwwiv

Co-authored-by: William Wernert <william.wernert@gmail.com>

JamesMConroy committed:
Renamed Colors to Attributes

As suggested by @rwwiv

Co-authored-by: William Wernert <william.wernert@gmail.com>

JamesMConroy committed:
Renamed colors to attributes

As suggested by rwwiv

Co-authored-by: William Wernert <william.wernert@gmail.com>

JamesMConroy committed:
Renamed colors to attributes

Also correctly used tput to assign blue color

As suggested by @rwwiv

Co-authored-by: William Wernert <william.wernert@gmail.com>

JamesMConroy committed:
Changed color variables to Attributes

As suggested by @rwwiv

Co-authored-by: William Wernert <william.wernert@gmail.com>

JamesMConroy committed:
Changed color variables to Attributes

As suggested by @rwwiv

Co-authored-by: William Wernert <william.wernert@gmail.com>

JamesMConroy committed:
Use spaces to define centerd justification output

As suggested by @rwwiv

Co-authored-by: William Wernert <william.wernert@gmail.com>

JamesMConroy committed:
Update the PADDING_CONSTENT to 15

As suggested by @rwwiv

Co-authored-by: William Wernert <william.wernert@gmail.com>

rwwiv reviewed (changes requested):
Approved this then caught another problem. Make sure you're updating this script through salt so you catch any jinja templating issues as well.

rwwiv commented:
You should also be sure to merge dev back into your branch since there have been changes to the file since you opened the PR.

JamesMConroy commented on salt/common/tools/sbin/so-status:
```diff
@@ -290,22 +287,32 @@ main() {
+    # don't print colors if NO_COLOR is set to anything
+    [ "${#NO_COLOR}" -ne 0 ] && __tty=0
+}
+
```
I don't usually work with salt. How do you update the script with it?

rwwiv replied:
Definitely don't remove stuff if you don't know its purpose, that's an easy way to break things. 

To answer your question about updating a script with salt: there is a master file for every script that's located in `/opt/so/saltstack/default/salt/common/tools/sbin` which the `common` salt state copies into `/usr/sbin` after running them through a tool called jinja, which replaces variables defined using its syntax. (Check `so-import-pcap` in the master directory for a good example of what I'm talking about re: jinja syntax.) For reference, the common state can be run using the following command:
```bash
sudo salt-call state.apply common
```

That being said, since we don't currently have much in the way of developer documentation I would do some reading on saltstack and take a closer look at the rest of the project to get a better idea of how Security Onion works as a whole. This project relies fairly heavily on salt and some minimal knowledge of its major features is important to contribute code that works as expected.

JamesMConroy committed:
Move the jinja endraw directive below is_tty

This will prevent jninja from interpreting the shell string length
expansion as the start of jninja comments

rwwiv commented on salt/common/tools/sbin/so-status:
```diff
@@ -101,9 +97,13 @@ create_expected_container_list() {
 }
 
```
```suggestion

# {% raw %}
```

rwwiv commented on salt/common/tools/sbin/so-status:
```diff
@@ -170,78 +170,73 @@ parse_status() {
 # {% raw %}
```
```suggestion
```

rwwiv reviewed (changes requested):
There were more issues with Jinja templating that I missed on my first look through your code. The suggested changes will get us to a place where your code can be merged.

JamesMConroy committed:
Update salt/common/tools/sbin/so-status

Add salt raw directive

Co-authored-by: William Wernert <william.wernert@gmail.com>

JamesMConroy committed:
Update salt/common/tools/sbin/so-status

Removed # {% raw %} from line 170

Co-authored-by: William Wernert <william.wernert@gmail.com>

rwwiv reviewed (approved):
Be sure in the future when submitting pull requests to submit functionality changes and refactoring separately, it make the review process much more difficult when you do both in a single PR.
